tools/metactl/src/snapshot.rs

```diff
@@ -0,0 +1,322 @@
+
+use std::collections::BTreeMap;
+use std::collections::BTreeSet;
+use std::fs::File;
+use std::io;
+use std::io::BufRead;
+use std::io::BufReader;
+use std::io::Lines;
+use std::io::Write;
+use std::net::SocketAddr;
+use std::str::FromStr;
+
+use anyhow::anyhow;
+use common_base::base::tokio;
+use common_meta_raft_store::config::RaftConfig;
+use common_meta_raft_store::log::RaftLog;
+use common_meta_raft_store::sled_key_spaces::KeySpaceKV;
+use common_meta_raft_store::state::RaftState;
+use common_meta_raft_store::state_machine::StateMachine;
+use common_meta_sled_store::get_sled_db;
+use common_meta_sled_store::init_sled_db;
+use common_meta_types::Cmd;
+use common_meta_types::Endpoint;
+use common_meta_types::LogEntry;
+use common_meta_types::LogId;
+use common_meta_types::Node;
+use databend_meta::export::deserialize_to_kv_variant;
+use databend_meta::export::serialize_kv_variant;
+use openraft::raft::Entry;
+use openraft::raft::EntryPayload;
+use openraft::Membership;
+use tokio::net::TcpSocket;
+use url::Url;
+
+use crate::export_meta;
+use crate::Config;
+
+pub async fn export_data(config: &Config) -> anyhow::Result<()> {
+    let raft_config = &config.raft_config;
+
+    // export from grpc api if metasrv is running
+    if config.grpc_api_address.is_empty() {
+        eprintln!("export meta dir from: {}", raft_config.raft_dir);
+
+        init_sled_db(raft_config.raft_dir.clone());
+        export_from_dir(config.db.clone())?;
+    } else {
+        export_from_running_node(&config).await?;
+    }
+
+    return Ok(());
+}
+
+pub async fn import_data(config: &Config) -> anyhow::Result<()> {
+    let raft_config = &config.raft_config;
+    eprintln!("import meta dir into: {}", raft_config.raft_dir);
+
+    init_sled_db(raft_config.raft_dir.clone());
+
+    clear()?;
+    let max_log_id = import_from(config.db.clone())?;
+
+    if config.initial_cluster.is_empty() {
+        return Ok(());
+    }
+    return init_new_cluster(config.initial_cluster.clone(), max_log_id).await;
+}
+
+// return the max log id
+fn import_lines<B: BufRead>(lines: Lines<B>) -> anyhow::Result<LogId> {
+    let db = get_sled_db();
+    let mut trees = BTreeMap::new();
+    let mut n = 0;
+    let mut max_log_id: LogId = LogId::default();
+    for line in lines {
+        let l = line?;
+        let (tree_name, kv_variant): (String, KeySpaceKV) = serde_json::from_str(&l)?;
+
+        // eprintln!("line: {}", l);
+
+        if !trees.contains_key(&tree_name) {
+            let tree = db.open_tree(&tree_name)?;
+            trees.insert(tree_name.clone(), tree);
+        }
+
+        let tree = trees.get(&tree_name).unwrap();
+
+        let (k, v) = serialize_kv_variant(&kv_variant)?;
+
+        tree.insert(k, v)?;
+        n += 1;
+
+        match kv_variant {
+            KeySpaceKV::Logs { key: _, value } => {
+                if value.log_id > max_log_id {
+                    max_log_id = value.log_id;
+                }
+            }
+            _ => {}
```

Comment:
There is a chance there is no log at all, e.g., when a snapshot is installed that overrides all logs.

In this case it should find out the max log id from state machine.